// ali-rds/lib/connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RDSConnection = void 0;
const node_util_1 = require("node:util");
const operator_1 = require("./operator");
const kWrapToRDS = Symbol('kWrapToRDS');
class RDSConnection extends operator_1.Operator {
    conn;
    constructor(conn) {
        super(conn);
        this.conn = conn;
        if (!this.conn[kWrapToRDS]) {
            [
                'query',
                'beginTransaction',
                'commit',
                'rollback',
            ].forEach(key => {
                this.conn[key] = (0, node_util_1.promisify)(this.conn[key]);
            });
            this.conn[kWrapToRDS] = true;
        }
    }
    release() {
        return this.conn.release();
    }
    async _query(sql) {
        return await this.conn.query(sql);
    }
    async beginTransaction() {
        return await this.conn.beginTransaction();
    }
    async commit() {
        return await this.conn.commit();
    }
    async rollback() {
        return await this.conn.rollback();
    }
}
exports.RDSConnection = RDSConnection;
